src/Symfony/Component/Messenger/MessageBus.php

```diff
@@ -11,28 +11,29 @@
 
 namespace Symfony\Component\Messenger;
 
-use Symfony\Component\Messenger\Exception\InvalidArgumentException;
 use Symfony\Component\Messenger\Middleware\MiddlewareInterface;
 
 /**
  * @author Samuel Roze <samuel.roze@gmail.com>
  * @author Matthias Noback <matthiasnoback@gmail.com>
+ * @author Nicolas Grekas <p@tchwork.com>
  */
 class MessageBus implements MessageBusInterface
 {
-    private $middlewareHandlers;
-
-    /**
-     * @var MiddlewareInterface[]|null
-     */
-    private $indexedMiddlewareHandlers;
+    private $middlewareAggregate;
 
     /**
      * @param MiddlewareInterface[]|iterable $middlewareHandlers
      */
     public function __construct(iterable $middlewareHandlers = array())
     {
-        $this->middlewareHandlers = $middlewareHandlers;
+        if ($middlewareHandlers instanceof \IteratorAggregate) {
+            $this->middlewareAggregate = $middlewareHandlers;
+        } elseif (\is_array($middlewareHandlers)) {
+            $this->middlewareAggregate = new \ArrayObject($middlewareHandlers);
+        } else {
+            $this->middlewareAggregate = new \ArrayObject(iterator_to_array($middlewareHandlers, false));
```

Comment:
previously the expansion (iterator_to_array) happened lazily, should we preserve it?